server/etcdserver/txn/txn_test.go

```diff
@@ -94,3 +98,429 @@ func TestWriteTxnPanic(t *testing.T) {
 
 	assert.Panics(t, func() { Txn(ctx, zaptest.NewLogger(t), txn, false, s, &lease.FakeLessor{}) }, "Expected panic in Txn with writes")
 }
+
+func TestCheckTxnAuth(t *testing.T) {
+	lg := zaptest.NewLogger(t)
+
+	be, _ := betesting.NewDefaultTmpBackend(t)
+	defer betesting.Close(t, be)
+
+	simpleTokenTTLDefault := 300 * time.Second
+	tokenTypeSimple := "simple"
+	dummyIndexWaiter := func(index uint64) <-chan struct{} {
+		ch := make(chan struct{}, 1)
+		go func() {
+			ch <- struct{}{}
+		}()
+		return ch
+	}
+
+	tp, _ := auth.NewTokenProvider(zaptest.NewLogger(t), tokenTypeSimple, dummyIndexWaiter, simpleTokenTTLDefault)
+
+	as := auth.NewAuthStore(lg, schema.NewAuthBackend(lg, be), tp, 4)
+
+	// create "root" user and "foo" user with limited range
+	if _, err := as.RoleAdd(&pb.AuthRoleAddRequest{Name: "root"}); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := as.RoleAdd(&pb.AuthRoleAddRequest{Name: "rw"}); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := as.RoleGrantPermission(&pb.AuthRoleGrantPermissionRequest{
+		Name: "rw",
+		Perm: &authpb.Permission{
+			PermType: authpb.READWRITE,
+			Key:      []byte("foo"),
+			RangeEnd: []byte("zoo"),
+		},
+	}); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := as.UserAdd(&pb.AuthUserAddRequest{Name: "root", Password: "foo"}); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := as.UserAdd(&pb.AuthUserAddRequest{Name: "foo", Password: "foo"}); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := as.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: "root", Role: "root"}); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := as.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: "foo", Role: "rw"}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := as.AuthEnable(); err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name       string
+		txnRequest *pb.TxnRequest
+		err        error
+	}{
+		{
+			name: "Out of range compare is unauthorized",
+			txnRequest: &pb.TxnRequest{
+				Compare: []*pb.Compare{
+					{
+						Key:      []byte("boo"),
+						RangeEnd: []byte("zoo"),
+					},
+				},
+				Success: []*pb.RequestOp{},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "In range compare is authorized",
+			txnRequest: &pb.TxnRequest{
+				Compare: []*pb.Compare{
+					{
+						Key:      []byte("foo"),
+						RangeEnd: []byte("zoo"),
+					},
+				},
+				Success: []*pb.RequestOp{},
+			},
+			err: nil,
+		},
+		{
+			name: "Nil request range is always authorized",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestRange{
+							RequestRange: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
```

Comment:
Those tests have needles repetition and bloat. Have you thought about refactoring some parts?

For example creating variable for repeating parts
```
inRangeRequest := RequestOp{
	Request: &pb.RequestOp_RequestRange{
		RequestRange: &pb.RangeRequest{
			Key:      []byte("foo"),
			RangeEnd: []byte("zoo"),
		},
	},
},
```
Then the test could be just
```
{
	name: "Range request in range is authorized",
	txnRequest: &pb.TxnRequest{
		Success: []*pb.RequestOp{inRangeRequest},
		Failure: []*pb.RequestOp{inRangeRequest},
	},
	err: nil,
},
```

Reply:
Makes sense.